lntest/itest/lnd_channel-funding_fund_max_test.go

```diff
@@ -0,0 +1,277 @@
+			SendAll: true,
+		}); err != nil {
+			ht.Fatalf("unable to sweep coins from node %v: %v",
+				node.Name(), err)
+		}
+
+		// Ensures we don't leave any transaction in the mempool after sweeping.
+		mineBlocks(ht, net, 1, 1)
+
+		// Make sure all funds were swept.
+		checkChannelBalance(node, 0)
+	}
+
+	runTestCase := func(carol, dave *lntest.HarnessNode,
+		testCase *chanFundMaxTestCase) func(t *testing.T) {
+
+		return func(t *testing.T) {
+			net.SendCoins(ht.t, testCase.amount, carol)
+			defer func() {
+				if testCase.amount <= 2000 {
+					// Add additional funds to sweep "dust" UTXO.
+					net.SendCoins(ht.t, 100000, carol)
+				}
+
+				// Remove all funds from Carol.
+				sweepNodeWalletAndAssert(carol)
+			}()
+
+			// The parameters to try opening the channel with.
+			chanParams := lntest.OpenChannelParams{
+				Amt:         0,
+				PushAmt:     testCase.pushAmt,
+				SatPerVByte: testCase.feeRate,
+				FundMax:     true,
+			}
+
+			// If we don't expect the channel opening to be
+			// successful, simply check for an error.
+			if testCase.chanOpenShouldFail {
+
+				_, err := net.OpenChannel(
+					carol, dave, chanParams,
+				)
+
+				if err == nil {
+					t.Errorf("should not open channel "+
+						"for: %v", testCase.name)
+				}
+
+				return
+			}
+
+			// Otherwise, if we expect to open a channel use the helper function.
+			chanPoint := openChannelAndAssert(
+				ht, net, carol, dave, chanParams,
+			)
+			defer func() {
+				// Close the channel between Carol and Dave, asserting
+				// that the channel has been properly closed on-chain.
+				closeChannelAndAssert(
+					ht, net, carol, chanPoint, false,
+				)
+			}()
+
+			cType, err := channelCommitType(carol, chanPoint)
+			if err != nil {
+				ht.Fatalf("unable to get channel type: %v", err)
+			}
+
+			// Carol's balance should be her amount subtracted by the
+			// commitment transaciton fee.
+			err = checkChannelBalance(
+				carol, testCase.carolBalance-calcStaticFee(cType, 0),
+			)
+			if err != nil {
+				t.Errorf("carol's balance is wrong: %v", err)
+			}
+
+			// Ensure Dave's balance within the channel is equal to the push amount.
+			err = checkChannelBalance(dave, testCase.pushAmt)
+			if err != nil {
+				t.Errorf("dave's balance is wrong: %v", err)
+			}
+		}
+	}
+
+	var testCases = []*chanFundMaxTestCase{
+		{
+			name:               "wallet amount is dust",
+			amount:             2000,
+			chanOpenShouldFail: true,
+		},
+
+		{
+			name:   "wallet amount < min chan size (~18000sat)",
+			amount: 18000,
+			// Using a feeRate of 1 sat/vByte ensures that we test for
+			// min chan size and not excessive fees.
+			feeRate:            1,
+			chanOpenShouldFail: true,
+		},
+
+		{
+			name:   "wallet amount > min chan size (37000sat)",
+			amount: 37000,
+			// The transaction fee to open the channel must be subtracted from
+			// Carol's balance (since wallet balance < max-chan-size).
+			carolBalance: btcutil.Amount(37000) - fundingFee(1, false),
```

Comment:
We should add a test with a low balance and the anchor reserve, just to make sure the interplay here is correct.

Reply:
Was added by @bjarnemagnussen, to be reviewed here: https://github.com/lightningnetwork/lnd/pull/6903/commits/f85113904bed65eb650f5dd3c5de8445b1c59586#diff-583407cd98e4e423efd8a6afe4acb5b67395da6ae79927b11f9e3b989f822313R142